### ISSUE TITLE
Check interface compatibility before deploy

### ISSUE THREAD

**Issue by eduardonunesp** [trufflesuite/truffle#1207]
## Issue

On the current implementation, if you have a contract with an abstract interface and the overwritten function is implemented, but with a different arity, for instance: override the function name but not the params correctly, it will compile (perhaps this is the Solc compiler), however during the deployment it will fail with an error that doesn't say anything related to the incorrect implementation of the interface.

### Error

```bash
Error encountered, bailing. Network state unknown. Review successful transactions manually.
Error: The contract code couldn't be stored, please check your gas amount.
```

## Steps to Reproduce

Implement the contracts as the example and try to deploy:

```solidity
pragma solidity 0.4.24;

interface AbstractContract {
    function supply(bool cool) external view returns (uint256);
}

contract MyContract is AbstractContract {
    uint256 total = 10;
    
    function supply() public view returns (uint256) {
        return total;
    }
}
```

## Expected Behavior

It should warn that the interface isn't right like Remix does:

![image](https://user-images.githubusercontent.com/582516/44426296-a1859900-a564-11e8-966c-d6735fe93fe3.png)

## Actual Results

An verbose an error that leads to misunderstanding of the real issue

```bash
Error encountered, bailing. Network state unknown. Review successful transactions manually.
Error: The contract code couldn't be stored, please check your gas amount.
```

## Environment

* Operating System: `macOS 10.13.6`
* Ethereum client: `ganache-cli 6.1.6`
* Truffle version (`truffle version`): `Truffle v4.1.13 (core: 4.1.13)`
* Solidity version: `Solidity v0.4.24 (solc-js)`
* node version (`node --version`): `v8.11.4`
* npm version (`npm --version`):  `6.2.0`

**Comment by cgewecke** [comment]
Hi @eduardonunesp, thanks for this issue. In the new beta of Truffle we're throwing an error for this case:
```
Error:  *** Deployment Failed ***

"MyContract" is an abstract contract or an interface and cannot be deployed
   * Hint: just import the contract into the '.sol' file that uses it.
```

But it needs to be improved to add the case you raise here, where you *thought* you implemented all the functions but made a small mistake somewhere. 

NB: There's a similar [issue](https://github.com/ethereum/remix/issues/665) at Remix suggesting that the error should identify at least one of the unimplemented functions. To do that we'd need to detect this case and do some AST processing on the contract in question.

**Comment by eduardonunesp** [comment]
@cgewecke thanks for the feedback, is possible to add another hint too? isn't the best case scenario, but to warn that the interface on `MyContract` and `AbstractContract` aren't right already helps to avoid a waste of time trying to fix `gas` issues :)

**Comment by eduardonunesp** [comment]
Thanks 👍
